### PR TITLE
Issue 786 mysql ssl support

### PR DESCRIPTION
**Pull request by p-alik**
This PR aims to add ssl connection support for MySQL backend, which is requested in the issue #786 .
At this stage it allow setting of mysql option `MYSQL_OPT_SSL_MODE` only.
The connection could be established  with `MYSQL_OPT_SSL_MODE` set to `SSL_MODE_DISABLED`, `SSL_MODE_PREFERRED` and `SSL_MODE_REQUIRED`.

Tested with
`DATABASE_URL='mysql://***?ssl_mode=(disabled|required|preferred)' cargo test --manifest-path diesel/Cargo.toml --features mysql --no-default-features mysql`

https://dev.mysql.com/doc/c-api/8.0/en/mysql-options.html
https://dev.mysql.com/doc/c-api/8.0/en/c-api-encrypted-connections.html

### CODE REVIEW

**Comment by p-alik** [comment]
I've to improve this PR for sake of Windows and I suppose to do that next week.

**Comment by p-alik** [comment]
Due to https://github.com/sgrif/mysqlclient-sys/issues/27 there is no way to set ssl-mode for MySQL connection on Windows in same way as on Linux/Unix yet.

Would it be appropriate to restrict setting of MySQL option `MYSQL_OPT_SSL_MODE` for `#[cfg(not(windows))]`? Any other suggestions?

**Comment by ELD** [comment]
That would probably be a suitable alternative for now, since the corresponding Windows bindings don’t work right now.

File: diesel/Cargo.toml
Comment (weiznich) [inline diesel/Cargo.toml:20]: `mysqlclient-sys 0.2.2` is the first version that supports the necessary functions, therefore I do not see any reason to bump the version here to 0.2.4.

File: diesel/src/mysql/connection/raw.rs
Comment (weiznich) [inline diesel/src/mysql/connection/raw.rs:192]: Is this a fundamental restriction of `libmysqlclient` that ssl is not supported on windows or is this a restriction of the `libmysqlclient-sys` crate?

File: diesel/src/mysql/connection/raw.rs
Comment (weiznich) [inline diesel/src/mysql/connection/raw.rs:54]: Can we make `ConnectionOptions::ssl_mode()` just unconditionally return `None` on windows instead of having `cfg`'s all over the code?

**Review by weiznich — COMMENTED** [review]
I've left some comments, otherwise this is looking good :+1:

(Sorry for taking some time to do a review here, it took some time to catch up with everything after holiday.)

File: diesel/src/mysql/connection/raw.rs
Comment (p-alik) [reply diesel/src/mysql/connection/raw.rs:192]: It is a restriction of `libmysqlclient-sys`. There is a PR https://github.com/sgrif/mysqlclient-sys/pull/28
which would make all `not(windows)` in this once checks obsolete.

Comment (weiznich) [reply diesel/src/mysql/connection/raw.rs:192]: I would rather not like to introduce such special case handling then. The current situation with mysqlclient seems to be quite unfortunate as I cannot merge PR's there. As we've hit that problem already more than once we should definitively seek a solution for this. Maybe @sgrif can grant me the corresponding rights or transfer the repositories to the diesel organization?

Beside of that as a long term goal it would probably be great to replace libmysqlclient with a pure rust implementation.

Comment (p-alik) [reply diesel/src/mysql/connection/raw.rs:192]: > Beside of that as a long term goal it would probably be great to replace libmysqlclient with a pure rust implementation.

You've mentioned [mysql](https://docs.rs/mysql/19.0.1/mysql/) crate in https://github.com/sgrif/mysqlclient-sys/issues/25#issuecomment-623016021 in regard to this matter. But is performance of the crate comparable to `libmysqlclient-sys`?

Comment (weiznich) [reply diesel/src/mysql/connection/raw.rs:192]: I'm not sure if the performance is comparable. That's one of the many open questions that need to be answered before such an implementation can be merged. As starting point have a look at the benchmarks I've pushed [here](https://github.com/weiznich/diesel/tree/more/benches/diesel_bench) (Improvements are welcome)

**Comment by biot023** [comment]
Hi -- I guess nothing happened with this?
Are there any workarounds for SSL-secured access?
I'm going to talk to our DBA service about maybe SSH tunnelling but I'm not confident that will work out...

**Comment by weiznich** [comment]
@biot023 Commenting on old issues/PR's just to ask for updates is something that is highly discouraged. Please stop doing that if you have nothing meaningful (like how to resolve the issues outlined above) to contribute.

**Comment by biot023** [comment]
> @biot023 Commenting on old issues/PR's just to ask for updates is something that is highly discouraged. Please stop doing that if you have nothing meaningful (like how to resolve the issues outlined above) to contribute.

It was a genuine question. Never mind, consider me suitably discouraged.

**Comment by biot023** [comment]
Maybe not entirely discouraged.
I've been able to set up ProxySQL (https://proxysql.com) in a docker instance to handle the SSL bit of the connection.
I can just point any `diesel` services at that image.
Hope that's of some help to somebody else struggling with this.

**Comment by Neo-Zhixing** [comment]
Maybe as a stopgap measure we should fork `libmysqlclient-sys` into the diesel org for now and merge all pending PRs there? There are no new commits in that repo since 2019 and @sgrif has been inactive since July 2021.  MySQL SSL support is a real use case, and doing that would unblock us on multiple issues.

**Comment by weiznich** [comment]
@Neo-Zhixing There is no need to fork mysqlclient-sys. I have all merge and publish rights for that crate. Unfortunately I do not have the time to do a lot active work there while maintaining diesel at the same time, so this is basically waiting for someone laying out a basic plan about the future of mysqlclient-sys and someone who implements it.

**Comment by thomasmost** [comment]
> I've been able to set up ProxySQL (https://proxysql.com) in a docker instance to handle the SSL bit of the connection.
> I can just point any diesel services at that image.
> Hope that's of some help to somebody else struggling with this.

@biot023 this is I think exactly what I'll need to do—you don't happen to have an example Dockerfile/diesel config that you'd feel comfortable sharing do you (scrubbed of anything sensitive, obviously)?

**Comment by biot023** [comment]
> @biot023 this is I think exactly what I'll need to do—you don't happen to have an example Dockerfile/diesel config that you'd feel comfortable sharing do you (scrubbed of anything sensitive, obviously)?

Hi -- I really struggled with this. I got it working, as I recall, but then had an issue with tons of connections being established to the main server and never dropped -- took the whole site down in about 4 minutes, when I tested it!
It's probable that you could get it working but in our case I went past the point of sunken cost fallacy and we abandoned diesel for that particular project.
That's not throwing shade on diesel, btw -- I still use it whenever I can, it just didn't work out in this scenario for us.
Sorry I'm not more help, man.
Cheers,
   Doug.

**Comment by weiznich** [comment]
@biot023  and @thomasmost Just to reiterate what I've written above: I would be more than happy to merge support for mysql ssl connections to diesel, but that requires work from someone. I'm happy to provide some pointers + discuss design, but implementation needs to be done by someone else.

**Comment by thomasmost** [comment]
@weiznich Let's do it! What is your schedule like/what forum would you prefer for such a discussion?

**Comment by weiznich** [comment]
@thomasmost I had a quick look at open questions here: The interface proposed in this PR looks good to me. There are the following open points from my point of view:

* [ ] Windows support
    * This is the largest open point
    * Requires changes to [`mysqlclient-sys`](https://github.com/sgrif/mysqlclient-sys), which may require restructuring the code there. That's the point that is quite fuzzy for me now.
* [ ] Rebasing the PR to the current master branch
* [ ] Tests?

**Comment by thomasmost** [comment]
Got it! I will see if I can get in there and work this out.

File: diesel/Cargo.toml
Comment (weiznich) [reply diesel/Cargo.toml:20]: We definitively need to update the minimal supported mysqlclient-sys version here to something newer than even released. My old comment was written with the assumption that the old release also contains all necessary functions. 

Generally speaking: We try not to update the minimal supported mysqlclient-sys version as that is considered a breaking change, due to the fact that it links a native library and cargo only allows one version of `mysqlclient-sys` in your dependency tree.

**Comment by p-alik** [comment]
Will rebase the feature branch.

File: diesel/Cargo.toml
Comment (p-alik) [reply diesel/Cargo.toml]: 04c4dc482d8cf3291213a74f810b9030078dd192 brings 0.2.2

File: diesel/src/mysql/connection/raw.rs
Comment (p-alik) [reply diesel/src/mysql/connection/raw.rs:54]: Is it convenient to do that this way?
3b9fd78aa9f878bb7552bb1ffa831c89365bfa82

**Comment by thomasmost** [comment]
p-alik how can I help get these tests passing?

File: diesel/src/mysql/connection/raw.rs
Comment (weiznich) [reply diesel/src/mysql/connection/raw.rs:193]: @thomasmost That's the most important open issue. Someone with access to a windows development environment needs to update mysqlclient-sys.

**Comment by thomasmost** [comment]
@weiznich now this is lookin pretty good

**Comment by weiznich** [comment]
@thomasmost Yes the diesel side of this change looks good, beside the fact that this only supports linux/macos. I would like to solve the windows issue first before finally merging this. As already mentioned that would require updating the bindings in mysqlclient-sys, which requires access to bindgen on windows. That's unfortunately something I cannot help with other than merging PR's and cutting releases.
I image something like following here:
- [ ] Update the generated bindings in mysqlclient-sys via a PR there (that should likely update bindings for both variants)
- [ ] Update this PR to use the mysqlclient-sys version based on the previous PR to check everything works fine
- [ ] I will merge the mysqlclient-sys PR + cut a release there
- [ ] Update this PR again to use the now newly released mysqlclient-sys version
- [ ] Merge this PR

**Comment by thomasmost** [comment]
So we are waiting on this PR then: https://github.com/sgrif/mysqlclient-sys/pull/28/files ?

**Comment by weiznich** [comment]
I've totally forgot about that PR 🙈  Merged it and cut a new [`mysqlclient-sys` release](https://crates.io/crates/mysqlclient-sys) containing the corresponding changes.

That means the following points are open before this PR can be merged:
- [ ] Update the PR to use the new `mysqlclient-sys` release + update the code to work on all platforms
- [ ] Add documentation about this new feature to `MysqlConnection::establish`
- [ ] Add a changelog entry.

As a warning note for @thomasmost: This PR targets the master branch, which contains quite a few backward incompatible  changes compared to the 1.4.x release. Additionally we won't immediately cut a new release after this PR is merged, just to be clear about that upfront.

**Comment by thomasmost** [comment]
Understood. Totally understand. Thanks for the update/attention to this @weiznich and @p-alik 🙏🏼

File: diesel/Cargo.toml
Comment (thomasmost) [inline diesel/Cargo.toml:20]:
```suggestion
mysqlclient-sys = { version = "0.2.5", optional = true }
```

File: diesel/src/mysql/connection/url.rs
Comment (thomasmost) [inline diesel/src/mysql/connection/url.rs:11]: all of these `cfg` flags can be retired now, yeah?

Comment (p-alik) [reply diesel/src/mysql/connection/url.rs]: yes.

**Review by weiznich — APPROVED** [review]
Looks good for me if the following things are added:

* A changelog entry [here](https://github.com/diesel-rs/diesel/blob/master/CHANGELOG.md#added) + adjusting the minimal supported mysqlclient-sys version under ["REMOVED"](https://github.com/diesel-rs/diesel/blob/master/CHANGELOG.md#removed)
* A note in the documentation of `MysqlConnection::establish` on the available options.

**Comment by p-alik** [comment]
Added requested notes to the changelog.

**Review by thomasmost — APPROVED** [review]
LGTM!

**Comment by thomasmost** [comment]
@weiznich one last follow-up; just so I understand, are we talking days, weeks, or months before this gets included in a release of the crate?

**Comment by weiznich** [comment]
> one last follow-up; just so I understand, are we talking days, weeks, or months before this gets included in a release of the crate?

We generally do not give any estimates when a certain release happens other than: When it's done. Additionally as this PR is not merged yet (due to missing documentation) it's not certain if the next release will contain this feature or not.

> Until that time, any reason I can't get my app working with an (admittedly, unstable) build using the master branch as its dependency, i.e.

Please open an discussion thread about that with more details about what you understand under "can't get working". A PR is definitively not the right place to discuss such issues.

**Comment by p-alik** [comment]
> as this PR is not merged yet (due to missing documentation) it's not certain if the next release will contain this feature or not.
24b612d08d2826627c98a17819aaef834feab711 aims to add missed documentation.

**Comment by thomasmost** [comment]
> > Until that time, any reason I can't get my app working with an (admittedly, unstable) build using the master branch as its dependency, i.e.
> 
> Please open an discussion thread about that with more details about what you understand under "can't get working". A PR is definitively not the right place to discuss such issues.


I think you misunderstood me @weiznich — I wasn't asking for support lol — nvm

Thanks again for your work on this crate

**Comment by yerke** [comment]
@p-alik @weiznich Thank you very much for working on this!

**Comment by ToniIltanen** [comment]
Im having problems with the ssl_mode=DISABLED, it seems diesel is not passing it to the actual connection, thus the connections ends up with `SSL connection error: error:0A000102:SSL routines::unsupported protocol`, no matter if the ssl_mode=DISABLED is added on the connection string.

**Comment by weiznich** [comment]
@ToniIltanen Please don't post the same issue in multiple places. You won't get an faster answer by doing that, you just burn time of other people by doing that. At least for me that signals that this issue can be moved down on my priority list.

If you really need a timely answer from this project you should consider a support contract otherwise you need to wait till someone has time to investigate your problem.

**Comment by ToniIltanen** [comment]
@weiznich I myself dont need this fixed, it was just an observation i made when i tested diesel with my clients db. I already fixed the issua by upgrading the db. But if you wish, I can remove the issue if that is a better solution for you.

**Comment by weiznich** [comment]
@ToniIltanen Again: That's no reason to post to an old semi-related issue. If you feel that this issue is related to your issue you can link it from your issue instead of posting here and pinging 7 other people as well. Stop doing that.

**Comment by libsilverwolf** [comment]
Is this feature also usable with `diesel-cli` ?

I tried

```
DATABASE_URL="mysql://***:3306/port?ssl_mode=required" diesel print-schema
```

but got 

```
Could not connect to database via `mysql://***:3306/port?ssl_mode=required`: Connections using insecure transport are prohibited while --require_secure_transport=ON.
```

I'm using MariaDB.

**Pull request merged**
